packages/lyra/src/prefix-tree/trie.ts

```diff
@@ -112,12 +112,12 @@ export class Trie {
     if (!word) return false;
 
     function removeWord(node: TrieNode, _word: string, docID: string): boolean {
-      const [nodeWord /**_docs*/] = node.getWord();
+      const [nodeWord, docIDs] = node.getWord();
 
       if (node.end || (exact && node.end && nodeWord === word)) {
         node.removeDoc(docID);
 
-        if (node.children?.size) {
+        if (node.children?.size && docIDs.has(docID)) {
```

Comment:
This fixes the issue, but for the wrong reason :)

`docIDs.has(docID)` will always return `false` here, because in the previous line `node.removeDoc(docID)` removed the document from the set. As a result, `node.end = false` is never executed, so the whole `if` block here could be completely removed.

Never setting `node.end = false` on one hand fixes the issue, and does not cause any correctness problem. It just leaves a term in the trie, even when all documents associated to it were removed. This is fine, but will have performance impacts on large tries if a lot of documents are deleted: even when the set of documents associated to a term is empty, the term is never removed, so it is still considered in prefix and fuzzy search.

My recommendation: change the `if` condition to something like `node.children?.size && docIDs.size === 0`. That means: if after deleting this one document there are no more documents in the set pointed by this word, remove the word from the trie by making the node non-terminal.

Note that this still leaves one case uncovered, namely when the node its terminal but has no children. In that case, in order to really delete the term from the trie, one has to backtrack to the closest branching, and adjust the nodes to remove the "orphaned" suffix. This could be addressed in another issue, since again it does not affect correctness, but rather only performance in case of a large trie with many deleted terms (that without the proposed refactoring stay in the trie, even though they point to empty document sets).

I am confident that @micheleriva understands what I am pointing at, but otherwise I can offer some more explanation.